### ISSUE TITLE
SAM3VideoSemanticPredictor silently loses all tracking with an incompatible SAM3 checkpoint — `_load_checkpoint` discards `strict=False` mismatches instead of surfacing them

### ISSUE THREAD

**Issue by amlan-nayak** [ultralytics/ultralytics#25152]
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and found no similar bug report.

### Ultralytics YOLO Component

Predict, Multi-GPU

### Bug

SAM3VideoSemanticPredictor works correctly with the standard sam3.pt checkpoint, but silently loses ALL cross-frame identity when loaded from a different, valid SAM3 checkpoint whose tracker weights use a `tracker.model.*` key prefix (I have one locally, sam3.1.pt, 3.50 GB, containing `detector.backbone.vision_backbone.interactive_convs.*` and `tracker.model.{maskmem_tpos_enc,interactivity_no_mem_embed,no_obj_embed_spatial,output_valid_embed,output_invalid_embed,...}` — architecturally a superset of sam3.pt, apparently an "interactive"/promptable SAM3 variant).

With this checkpoint, every frame gets a brand-new, non-overlapping block of track ids and this was confirmed even on a video made of the same frame repeated 6 times, where a working tracker should trivially reuse the same ids:

0 n= 5 ids= [0.0, 1.0, 2.0, 3.0, 4.0]
1 n= 5 ids= [5.0, 6.0, 7.0, 8.0, 9.0]
2 n= 5 ids= [10.0, 11.0, 12.0, 13.0, 14.0]
...

Swapping only the checkpoint path to sam3.pt, everything else identical, gives the correct result ([0,1,2,3,4] every frame). This is 100% reproducible in both directions.

### Root cause

ultralytics/models/sam/build_sam3.py::_load_checkpoint remaps checkpoint keys with hardcoded string substitutions before a non-strict load:

def _load_checkpoint(model, checkpoint, interactive=False):
    ...
    if interactive:
        ...
        sam3_image_ckpt.update({k.replace("tracker.transformer.encoder", "memory_attention"): v for k, v in ckpt.items() if "tracker.transformer" in k})
        sam3_image_ckpt.update({k.replace("tracker.maskmem_backbone", "memory_encoder"): v for k, v in ckpt.items() if "tracker.maskmem_backbone" in k})
        sam3_image_ckpt.update({k.replace("tracker.", ""): v for k, v in ckpt.items() if "tracker." in k})
    model.load_state_dict(sam3_image_ckpt, strict=False)  # <-- return value discarded, no warning on mismatch
    return model

My checkpoint's tracker weights are prefixed `tracker.model.*`, not `tracker.maskmem_backbone.*` / `tracker.transformer.encoder.*`, so neither specific rename rule fires. The generic fallback (`tracker.` → "") turns e.g. `tracker.model.maskmem_tpos_enc` into `model.maskmem_tpos_enc`, which doesn't match any real parameter name on the freshly-constructed SAM3Model (there's no top-level `model.` submodule), so it's silently dropped as an "unexpected key," and the real target parameter is left at its random initialization. Because load_state_dict(..., strict=False) is called without capturing/inspecting the returned _IncompatibleKeys, none of this surfaces as an error or even a log line, regardless of verbose.

I instrumented this directly. Loading sam3.pt:
missing_keys count: 0
unexpected_keys count: 1710  (all backbone.* -- expected, since with_backbone=False)

Loading my sam3.1.pt-style checkpoint:
missing_keys count: 309
missing_keys sample: ['maskmem_tpos_enc', 'no_mem_embed', 'no_mem_pos_enc', 'no_obj_ptr', 'no_obj_embed_spatial']
unexpected_keys count: 2097
unexpected_keys sample: ['backbone.vision_backbone.interactive_convs.0.dconv_2x2_0.weight', ...]

Those 309 missing keys are exactly the memory-encoder/attention parameters SAM2-style video tracking depends on. With them randomly initialized, I traced the downstream effect frame-by-frame: _associate_det_trk correctly matches detections to tracks on frame 0→1 (IoU ≈ 1.0), but by frame 1→2 the propagated track masks (tracker_low_res_masks_global) collapse into a near-identical, oversized blob for every object (in a ~79-object test, every masklet converged to ~8,080 px regardless of the real object's shape/size), IoU against real detections drops to ~0.01, nothing matches assoc_iou_thresh, and every detection is re-spawned as a new track from then on — which is the "new ids every frame" symptom above.

### Environment

Ultralytics 8.4.79 🚀 Python-3.11.6 torch-2.5.1 CUDA:0 (NVIDIA RTX A6000, 48541MiB)
Setup complete ✅ (128 CPUs, 1007.0 GB RAM, 103.0/435.5 GB disk)

OS                     Linux-5.14.0-687.12.1.el9_8.x86_64-x86_64-with-glibc2.34
Environment            Linux
Python                 3.11.6
Install                pip
RAM                    1007.02 GB
CPU                    Intel Xeon Gold 6338 CPU @ 2.00GHz
GPU                    NVIDIA RTX A6000, 48541MiB
CUDA                   12.4
torch                  2.5.1
torchvision            0.20.1

### Minimal Reproducible Example

The id-reset symptom itself only needs the standard, public sam3.pt — but it will NOT reproduce with that checkpoint (I confirmed sam3.pt tracks correctly). To reproduce the bug you need a checkpoint whose tracker weights use the tracker.model.* prefix; I can't attach mine, but the key-inspection snippet below reproduces the root cause against any checkpoint and shows the exact prefix mismatch:

import torch
from ultralytics.models.sam.build_sam3 import build_interactive_sam3
import ultralytics.models.sam.build_sam3 as b

checkpoint_path = "path/to/checkpoint.pt"  # any SAM3 checkpoint

captured = {}
orig = b._load_checkpoint
def patched(model, checkpoint, interactive=False):
    with open(checkpoint, "rb") as f:
        ckpt = torch.load(f, weights_only=False)
    if "model" in ckpt and isinstance(ckpt["model"], dict):
        ckpt = ckpt["model"]
    sam3_image_ckpt = {k.replace("detector.", ""): v for k, v in ckpt.items() if "detector" in k}
    sam3_image_ckpt.update({k.replace("backbone.vision_backbone", "image_encoder.vision_backbone"): v for k, v in sam3_image_ckpt.items() if "backbone.vision_backbone" in k})
    sam3_image_ckpt.update({k.replace("tracker.transformer.encoder", "memory_attention"): v for k, v in ckpt.items() if "tracker.transformer" in k})
    sam3_image_ckpt.update({k.replace("tracker.maskmem_backbone", "memory_encoder"): v for k, v in ckpt.items() if "tracker.maskmem_backbone" in k})
    sam3_image_ckpt.update({k.replace("tracker.", ""): v for k, v in ckpt.items() if "tracker." in k})
    res = model.load_state_dict(sam3_image_ckpt, strict=False)
    captured["res"] = res
    return model
b._load_checkpoint = patched
build_interactive_sam3(checkpoint_path, with_backbone=False)
b._load_checkpoint = orig

res = captured["res"]
print("missing_keys:", len(res.missing_keys), res.missing_keys[:10])
print("unexpected_keys:", len(res.unexpected_keys), res.unexpected_keys[:10])

Expected (any correctly-matched checkpoint): missing_keys: 0.
Actual (my tracker.model.*-prefixed checkpoint): missing_keys: 309, including maskmem_tpos_enc, no_mem_embed, no_mem_pos_enc, no_obj_ptr, no_obj_embed_spatial.

This is not just a naming mismatch, so a simple rename rule will not fix it. I tried adding an explicit tracker.model. → "" rename rule (mirroring the existing tracker.maskmem_backbone / tracker.transformer.encoder rules) and it fails immediately with shape mismatches, e.g.:

size mismatch for maskmem_tpos_enc: checkpoint [7,1,1,256] vs model [7,1,1,64]
size mismatch for no_obj_embed_spatial: checkpoint [16,256] vs model [1,64]
size mismatch for sam_mask_decoder.iou_token.weight: checkpoint [16,256] vs model [1,256]
size mismatch for sam_mask_decoder.mask_tokens.weight: checkpoint [48,256] vs model [4,256]
size mismatch for obj_ptr_tpos_proj.weight: checkpoint [256,256] vs model [64,256]

So this checkpoint's tracker submodule is a structurally different, larger configuration (256-dim memory features vs 64, 16 mask/object tokens vs 1) than what build_interactive_sam3() hardcodes (MemoryEncoder(out_dim=64, ...) and friends). Properly supporting it needs a second construction path with the right hyperparameters for this checkpoint family, which I can't responsibly guess at without knowing this variant's real spec — so I'm not proposing a fix here, just the precise diagnosis. Regardless of whether/how this checkpoint variant gets supported, the narrower and clearly-safe fix is: _load_checkpoint should stop discarding load_state_dict's return value, and should raise or log loudly whenever a strict=False load leaves any tracker-critical key missing or mismatched, instead of silently continuing with a partially-random model. That alone would have turned my multi-hour "why is tracking broken" investigation into an immediate, actionable error at load time.

### Additional

For anyone hitting the "brand new ids every frame" symptom: first check which checkpoint you're loading. If you have more than one SAM3 .pt file around, verify missing_keys is empty using the snippet above before assuming the tracker itself is broken — with a correctly-matched checkpoint (sam3.pt in my case) SAM3VideoSemanticPredictor tracks reliably.

### Are you willing to submit a PR?

- [x] Yes I'd like to help by submitting a PR!

**Comment by UltralyticsAssistant** [comment]
👋 Hello @amlan-nayak, thank you for the detailed SAM3 tracking report and diagnostics 🚀! The checkpoint comparison, `_load_checkpoint` instrumentation, environment details, and proposed PR are especially helpful.

We recommend reviewing the [Docs](https://docs.ultralytics.com/) for additional [Python](https://docs.ultralytics.com/usage/python) and [CLI](https://docs.ultralytics.com/usage/cli) usage examples. Please also confirm the behavior against the latest `ultralytics` package and include any relevant logs or test results with your report.

If this is a 🐛 Bug Report, please provide or maintain a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example). Since the reported behavior depends on a checkpoint with `tracker.model.*` keys that cannot be attached, please include a sanitized checkpoint key listing or a small fixture reproducing the incompatible load where possible.

For custom training ❓ questions, please provide dataset image examples and training logs, and verify that you are following our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips).

Join the Ultralytics community where it suits you best. For real-time chat, head to [Discord](https://discord.com/invite/ultralytics) 🎧. Prefer in-depth discussions? Check out [Discourse](https://community.ultralytics.com/). Or dive into threads on our [Subreddit](https://www.reddit.com/r/Ultralytics/) to share knowledge with the community.

This is an automated response 🤖; an Ultralytics engineer will also assist with reviewing the checkpoint compatibility and loading behavior soon.

## Upgrade

Upgrade to the latest `ultralytics` package including all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) in a [**Python>=3.8**](https://www.python.org/) environment with [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/) to verify your issue is not already resolved in the latest version:

```bash
pip install -U ultralytics
```

## Environments

YOLO may be run in any of the following up-to-date verified environments (with all dependencies including [CUDA](https://developer.nvidia.com/cuda)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/) and [PyTorch](https://pytorch.org/) preinstalled):

- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo26"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing. CI tests verify correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes) and [Tasks](https://docs.ultralytics.com/tasks) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

**Comment by Y-T-G** [comment]
Hello.

Thanks for reporting.

Since we do not claim to support unofficial checkpoints, it should be expected that these checkpoints may not work as intended. This isn't a bug because a bug is when something that's supposed to work isn't working as intended. The pipeline is designed and documented to work with the official SAM3 checkpoints.

**Comment by amlan-nayak** [comment]
@Y-T-G The issue was when i used the official SAM 3.1 checkpoints. The predictor runs without giving any warnings, and outputs incorrect tracks. Adding a check to show that only SAM 3.0 checkpoints work with the SAM3VideoSemanticPredictor might be useful

**Comment by Y-T-G** [comment]
I see. We do not currently have support for SAM3.1. It's not just the checkpoint structure that has changed with it, it also requires a different type of pipeline to make use of the enhancements it comes with for tracking.
